File: context/tools/confluence.py
"""Lee Confluence desde la terminal. SOLO LECTURA — no hay ningún verbo que escriba.

Existe porque la documentación de negocio (política de riesgo, contratos con lenders, PRDs) no está
en el código y el árbol no la puede indexar: vive en Confluence. Esto la baja a texto para poder
contrastarla contra el código, que es el único protocolo que la deja entrar al árbol.

    python3 tools/confluence.py espacios
    python3 tools/confluence.py paginas Creditop
    python3 tools/confluence.py leer 143786000
    python3 tools/confluence.py buscar "cupo rotativo"

Credenciales en `context/.env` (gitignoreado): CONFLUENCE_URL · CONFLUENCE_EMAIL · CONFLUENCE_TOKEN.

⚠ Lo que sale de acá NO es verdad todavía. Un documento desactualizado es indistinguible de uno
equivocado, así que toda afirmación se marca `confirmada` (el código coincide) / `contradicha` (difieren
— las más valiosas) / `no verificable` (política pura), y sólo las dos primeras entran a un nodo.
"""
import base64
import html
import json
import os
import re
import sys
import urllib.error
import urllib.parse
import urllib.request

RAIZ = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))


def credenciales():
    env = os.path.join(RAIZ, ".env")
    vals = dict(os.environ)
    if os.path.exists(env):
        with open(env, encoding="utf-8") as fh:
            for linea in fh:
                linea = linea.strip()
                if not linea or linea.startswith("#") or "=" not in linea:
                    continue
                k, v = linea.split("=", 1)
                vals.setdefault(k.strip(), v.strip().strip('"').strip("'"))
    faltan = [k for k in ("CONFLUENCE_URL", "CONFLUENCE_EMAIL", "CONFLUENCE_TOKEN") if not vals.get(k)]
    if faltan:
        sys.exit(f"faltan en context/.env: {', '.join(faltan)}")
    return vals["CONFLUENCE_URL"].rstrip("/"), vals["CONFLUENCE_EMAIL"], vals["CONFLUENCE_TOKEN"]


def get(ruta, params=None):
    base, email, token = credenciales()
    url = base + ruta + ("?" + urllib.parse.urlencode(params) if params else "")
    auth = base64.b64encode(f"{email}:{token}".encode()).decode()
    req = urllib.request.Request(url, headers={"Authorization": f"Basic {auth}", "Accept": "application/json"})
    try:
        with urllib.request.urlopen(req, timeout=60) as r:
            return json.load(r)
    except urllib.error.HTTPError as e:
        cuerpo = e.read()[:300].decode("utf-8", "replace")
        # ⚠ Atlassian NO contesta 401 cuando la credencial no sirve: en `/wiki` devuelve 403 y en la
        # API v2 devuelve **404**. Leído tal cual, un 404 en `/wiki/api/v2/spaces` se diagnostica como
        # ruta mal armada o espacio inexistente — y son tres pasos hasta descubrir que el token venció.
        # El que sí dice la verdad es el endpoint de Jira: 401 «Client must be authenticated».
        if e.code in (401, 403, 404):
            base, email, _ = credenciales()
            req = urllib.request.Request(
                base + "/rest/api/3/myself",
                headers={"Authorization": req.headers["Authorization"], "Accept": "application/json"},
            )
            try:
                urllib.request.urlopen(req, timeout=30).read()
            except urllib.error.HTTPError as sonda:
                if sonda.code == 401:
                    sys.exit(
                        f"HTTP {e.code} en {ruta}, pero la causa es la CREDENCIAL: "
                        f"{base}/rest/api/3/myself devuelve 401.\n"
                        f"El token de {email} venció o fue revocado. Generá uno nuevo en\n"
                        f"  https://id.atlassian.com/manage-profile/security/api-tokens\n"
                        f"y actualizá CONFLUENCE_TOKEN en context/.env (gitignoreado)."
                    )
            except Exception:
                pass  # la sonda es un extra: si falla, seguimos con el error original
        sys.exit(f"HTTP {e.code} en {ruta}: {cuerpo}")


# ── storage format (XHTML de Confluence) → texto plano legible ───────────────────────────────────
def a_texto(xhtml):
    t = xhtml or ""
    # macros que traen código o texto suelto
    t = re.sub(r'<ac:structured-macro[^>]*ac:name="code".*?<!\[CDATA\[(.*?)\]\]>.*?</ac:structured-macro>',
               r"\n```\n\1\n```\n", t, flags=re.S)
    t = re.sub(r"<ac:parameter[^>]*>.*?</ac:parameter>", "", t, flags=re.S)
    t = re.sub(r"<ri:page[^>]*ri:content-title=\"([^\"]*)\"[^>]*/?>", r"[[\1]]", t)
    t = re.sub(r"<ri:(user|attachment|url)[^>]*/?>", "", t)
    t = re.sub(r"</?ac:(link|inline-comment-marker|placeholder|adf-[a-z-]+)[^>]*>", "", t)
    t = re.sub(r"<ac:structured-macro[^>]*ac:name=\"([a-z-]+)\"[^>]*/>", r"«macro \1»", t)
    t = re.sub(r"</?ac:[a-z-]+[^>]*>", "", t)

    t = re.sub(r"<h1[^>]*>", "\n\n# ", t)
    t = re.sub(r"<h2[^>]*>", "\n\n## ", t)
    t = re.sub(r"<h3[^>]*>", "\n\n### ", t)
    t = re.sub(r"<h[4-6][^>]*>", "\n\n#### ", t)
    t = re.sub(r"</h[1-6]>", "\n", t)
    t = re.sub(r"<li[^>]*>", "\n- ", t)
    t = re.sub(r"</(p|div|tr|li|ul|ol|table)>", "\n", t)
    t = re.sub(r"<br\s*/?>", "\n", t)
    t = re.sub(r"</t[hd]>", " | ", t)
    t = re.sub(r"<[^>]+>", "", t)
    t = html.unescape(t)
    t = re.sub(r"[ \t]+\n", "\n", t)
    t = re.sub(r"\n{3,}", "\n\n", t)
    return t.strip()


def cmd_espacios():
    d = get("/wiki/api/v2/spaces", {"limit": 250})
    for s in sorted(d.get("results", []), key=lambda s: s.get("key", "")):
        if s.get("key", "").startswith("~"):
            continue  # espacios personales
        print(f"{s['id']:<12} {s.get('key',''):<16} {s.get('name','')}")


def cmd_paginas(clave):
    d = get("/wiki/api/v2/spaces", {"keys": clave})
    if not d.get("results"):
        sys.exit(f"no existe el espacio {clave}")
    sid = d["results"][0]["id"]
    cursor, total = None, 0
    while True:
        p = {"limit": 250}
        if cursor:
            p["cursor"] = cursor
        d = get(f"/wiki/api/v2/spaces/{sid}/pages", p)
        for pg in d.get("results", []):
            print(f"{pg['id']:<12} {pg['title']}")
            total += 1
        nxt = (d.get("_links") or {}).get("next")
        if not nxt:
            break
        q = urllib.parse.parse_qs(urllib.parse.urlparse(nxt).query)
        cursor = q.get("cursor", [None])[0]
        if not cursor:
            break
    print(f"\n— {total} páginas en {clave}", file=sys.stderr)


def cmd_leer(pid):
    d = get(f"/wiki/api/v2/pages/{pid}", {"body-format": "storage"})
    ver = d.get("version") or {}
    print(f"# {d.get('title','')}")
    print(f"<!-- id {pid} · v{ver.get('number','?')} · {ver.get('createdAt','?')} -->\n")
    print(a_texto(((d.get("body") or {}).get("storage") or {}).get("value", "")))


def cmd_buscar(texto):
    cql = f'text ~ "{texto}" and type = page'
    d = get("/wiki/rest/api/search", {"cql": cql, "limit": 40})
    for r in d.get("results", []):
        c = r.get("content") or {}
        esp = ((r.get("resultGlobalContainer") or {}).get("title")) or ""
        print(f"{c.get('id',''):<12} {esp:<20} {r.get('title','')}")


if __name__ == "__main__":
    if len(sys.argv) < 2:
        sys.exit(__doc__)
    cmd, args = sys.argv[1], sys.argv[2:]
    if cmd == "espacios":
        cmd_espacios()
    elif cmd == "paginas" and args:
        cmd_paginas(args[0])
    elif cmd == "leer" and args:
        cmd_leer(args[0])
    elif cmd == "buscar" and args:
        cmd_buscar(" ".join(args))
    else:
        sys.exit(__doc__)

File: connectors/atlassian/confluence.go
package atlassian

// La mitad Confluence del conector: la documentación de NEGOCIO (política de riesgo, contratos con
// lenders, PRDs), que no está en el código. SOLO LECTURA — no hay ningún método que escriba.
//
// ⚠ Lo que sale de acá NO es verdad todavía. Un documento desactualizado es indistinguible de uno
// equivocado, así que toda afirmación se contrasta contra el código antes de usarse.
//
// Hasta el 2026-09-24 esto vivía en Python (`tools/confluence.py`) con su propio cliente HTTP y su propio
// token, que había vencido sin que nada lo dijera; el de Jira, que es el mismo sitio y la misma cuenta,
// servía para los dos.

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"html"
	"net/http"
	"net/url"
	"regexp"
	"sort"
	"strings"

	"creditop/playground/lib/text"
)

// flexID acepta un id que Atlassian a veces manda como número y a veces como texto.
type flexID string

func (f *flexID) UnmarshalJSON(b []byte) error {
	var s string
	if err := json.Unmarshal(b, &s); err == nil {
		*f = flexID(s)
		return nil
	}
	var n json.Number
	if err := json.Unmarshal(b, &n); err != nil {
		return err
	}
	*f = flexID(n.String())
	return nil
}

// Space es un espacio de Confluence.
type Space struct {
	ID   flexID `json:"id"`
	Key  string `json:"key"`
	Name string `json:"name"`
}

// PageRef es una página en un listado.
type PageRef struct {
	ID    flexID `json:"id"`
	Title string `json:"title"`
}

// Page es una página con su cuerpo en el formato de almacenamiento (XHTML de Confluence).
type Page struct {
	Title   string `json:"title"`
	Version struct {
		Number    json.Number `json:"number"`
		CreatedAt string      `json:"createdAt"`
	} `json:"version"`
	Body struct {
		Storage struct {
			Value string `json:"value"`
		} `json:"storage"`
	} `json:"body"`
}

// SearchHit es un resultado de la búsqueda CQL.
type SearchHit struct {
	Title   string `json:"title"`
	Content struct {
		ID flexID `json:"id"`
	} `json:"content"`
	Container struct {
		Title string `json:"title"`
	} `json:"resultGlobalContainer"`
}

// CredentialError dice que la causa de un 403/404 de Confluence es la credencial, no la ruta.
type CredentialError struct {
	Status     int
	Path, Site string
	Email      string
}

func (e *CredentialError) Error() string {
	return fmt.Sprintf("HTTP %d en %s, pero la causa es la CREDENCIAL: %s/rest/api/3/myself devuelve 401.\n"+
		"El token de %s venció o fue revocado. Generá uno nuevo en\n"+
		"  https://id.atlassian.com/manage-profile/security/api-tokens\n"+
		"y actualizá ATLASSIAN_API_TOKEN en connectors/.env (gitignoreado).", e.Status, e.Path, e.Site, e.Email)
}

// wiki hace un GET de lectura y, si falla, dice por qué. ⚠ Atlassian NO contesta 401 cuando la
// credencial no sirve: en `/wiki` devuelve 403 y en la API v2 devuelve **404**. Leído tal cual, un 404
// en `/wiki/api/v2/spaces` se diagnostica como ruta mal armada o espacio inexistente — y son tres pasos
// hasta descubrir que el token venció. El que sí dice la verdad es el endpoint de Jira: 401.
func (c *Client) wiki(ctx context.Context, path string, params url.Values, out any) error {
	full := path
	if len(params) > 0 {
		full += "?" + params.Encode()
	}
	err := c.do(ctx, http.MethodGet, full, nil, out)
	var he *HTTPError
	if err == nil || !errors.As(err, &he) {
		return err
	}
	if he.Status == 401 || he.Status == 403 || he.Status == 404 {
		_, probe := c.GetMyself(ctx)
		var pe *HTTPError
		if errors.As(probe, &pe) && pe.Status == 401 {
			return &CredentialError{Status: he.Status, Path: path, Site: c.baseURL, Email: c.email}
		}
	}
	return fmt.Errorf("HTTP %d en %s: %s", he.Status, path, clip(he.Body, 300))
}

func clip(s string, n int) string {
	if len(s) > n {
		return s[:n]
	}
	return s
}

// Spaces lista los espacios, sin los personales (`~…`), ordenados por clave.
func (c *Client) Spaces(ctx context.Context) ([]Space, error) {
	var out struct {
		Results []Space `json:"results"`
	}
	if err := c.wiki(ctx, "/wiki/api/v2/spaces", url.Values{"limit": {"250"}}, &out); err != nil {
		return nil, err
	}
	var spaces []Space
	for _, s := range out.Results {
		if !strings.HasPrefix(s.Key, "~") {
			spaces = append(spaces, s)
		}
	}
	sort.SliceStable(spaces, func(i, j int) bool { return spaces[i].Key < spaces[j].Key })
	return spaces, nil
}

// ErrNoSpace: la clave no corresponde a ningún espacio.
var ErrNoSpace = errors.New("no existe el espacio")

// Pages lista todas las páginas de un espacio, siguiendo el cursor.
func (c *Client) Pages(ctx context.Context, key string) ([]PageRef, error) {
	var found struct {
		Results []Space `json:"results"`
	}
	if err := c.wiki(ctx, "/wiki/api/v2/spaces", url.Values{"keys": {key}}, &found); err != nil {
		return nil, err
	}
	if len(found.Results) == 0 {
		return nil, fmt.Errorf("%w %s", ErrNoSpace, key)
	}
	sid := string(found.Results[0].ID)
	var all []PageRef
	cursor := ""
	for {
		params := url.Values{"limit": {"250"}}
		if cursor != "" {
			params.Set("cursor", cursor)
		}
		var page struct {
			Results []PageRef `json:"results"`
			Links   struct {
				Next string `json:"next"`
			} `json:"_links"`
		}
		if err := c.wiki(ctx, "/wiki/api/v2/spaces/"+sid+"/pages", params, &page); err != nil {
			return nil, err
		}
		all = append(all, page.Results...)
		if page.Links.Next == "" {
			break
		}
		next, err := url.Parse(page.Links.Next)
		if err != nil {
			break
		}
		if cursor = next.Query().Get("cursor"); cursor == "" {
			break
		}
	}
	return all, nil
}

// ReadPage trae una página con su cuerpo.
func (c *Client) ReadPage(ctx context.Context, id string) (*Page, error) {
	var p Page
	if err := c.wiki(ctx, "/wiki/api/v2/pages/"+url.PathEscape(id), url.Values{"body-format": {"storage"}}, &p); err != nil {
		return nil, err
	}
	return &p, nil
}

// Search busca páginas por texto (CQL `text ~ "…" and type = page`), hasta 40.
func (c *Client) Search(ctx context.Context, query string) ([]SearchHit, error) {
	cql := fmt.Sprintf(`text ~ "%s" and type = page`, query)
	var out struct {
		Results []SearchHit `json:"results"`
	}
	if err := c.wiki(ctx, "/wiki/rest/api/search", url.Values{"cql": {cql}, "limit": {"40"}}, &out); err != nil {
		return nil, err
	}
	return out.Results, nil
}

// ── formato de almacenamiento (XHTML de Confluence) → texto plano legible ─────────────────────────

type rule struct {
	re   *regexp.Regexp
	with string
}

var storageRules = []rule{
	// macros que traen código o texto suelto
	{regexp.MustCompile(`(?s)<ac:structured-macro[^>]*ac:name="code".*?<!\[CDATA\[(.*?)\]\]>.*?</ac:structured-macro>`), "\n```\n${1}\n```\n"},
	{regexp.MustCompile(`(?s)<ac:parameter[^>]*>.*?</ac:parameter>`), ""},
	{regexp.MustCompile(`<ri:page[^>]*ri:content-title="([^"]*)"[^>]*/?>`), "[[${1}]]"},
	{regexp.MustCompile(`<ri:(user|attachment|url)[^>]*/?>`), ""},
	{regexp.MustCompile(`</?ac:(link|inline-comment-marker|placeholder|adf-[a-z-]+)[^>]*>`), ""},
	{regexp.MustCompile(`<ac:structured-macro[^>]*ac:name="([a-z-]+)"[^>]*/>`), "«macro ${1}»"},
	{regexp.MustCompile(`</?ac:[a-z-]+[^>]*>`), ""},
	{regexp.MustCompile(`<h1[^>]*>`), "\n\n# "},
	{regexp.MustCompile(`<h2[^>]*>`), "\n\n## "},
	{regexp.MustCompile(`<h3[^>]*>`), "\n\n### "},
	{regexp.MustCompile(`<h[4-6][^>]*>`), "\n\n#### "},
	{regexp.MustCompile(`</h[1-6]>`), "\n"},
	{regexp.MustCompile(`<li[^>]*>`), "\n- "},
	{regexp.MustCompile(`</(p|div|tr|li|ul|ol|table)>`), "\n"},
	{regexp.MustCompile(`<br\s*/?>`), "\n"},
	{regexp.MustCompile(`</t[hd]>`), " | "},
	{regexp.MustCompile(`<[^>]+>`), ""},
}

var (
	trailingBlanks = regexp.MustCompile(`[ \t]+\n`)
	manyNewlines   = regexp.MustCompile(`\n{3,}`)
)

// StorageToText pasa el cuerpo de una página a texto: encabezados como Markdown, listas con guion,
// celdas separadas por `|`, macros de código como bloque y enlaces a páginas como [[título]].
func StorageToText(xhtml string) string {
	t := xhtml
	for _, r := range storageRules {
		t = r.re.ReplaceAllString(t, r.with)
	}
	t = html.UnescapeString(t)
	t = trailingBlanks.ReplaceAllString(t, "\n")
	t = manyNewlines.ReplaceAllString(t, "\n\n")
	return strings.TrimFunc(t, text.IsSpace)
}
